src/br/com/ianffcs/apache_http_client_ring_adapter/client.clj

```diff
@@ -4,101 +4,185 @@
-              (getStatusCode [_this] status)
-              (getProtocolVersion [_this] protocol)))
-          (headerIterator [_this]
-            (let [headers (atom (for [[k vs] headers
-                                      v (if (coll? vs)
-                                          vs
-                                          [vs])]
-                                  (reify Header
-                                    (getName [_this] k)
-                                    (getValue [_this] (str v)))))]
-              (reify HeaderIterator
-                (hasNext [_this] (not (empty? @headers)))
-                (next [_this]
-                  (ffirst (swap-vals! headers rest)))))))))))
+  [{:keys [ring-handler
+           retry-fn
+           max-retries]
+    :or {max-retries 3}
+    :as   http-mock-map}]
+  (let [^ServiceUnavailableRetryStrategy
+        retry-strategy (when retry-fn
+                         (->service-unavailable-retry-strategy http-mock-map))
+        http-client-executed (proxy [CloseableHttpClient] []
+                               (getConnectionManager [_] nil)
+                               (close [_])
+                               (doExecute [^HttpHost target
+                                           ^HttpRequest http-request
+                                           ^HttpContext context]
+                                 (let [request-line (.getRequestLine http-request)
+                                       method (-> request-line
+                                                  .getMethod
+                                                  string/lower-case
+                                                  keyword)
+                                       request-uri (some-> request-line
+                                                           .getUri
+                                                           URI/create)
+                                       protocol (.getProtocolVersion request-line)
+                                       port (.getPort target)
+                                       headers (reduce (fn [headers
+                                                            ^Header header]
+                                                         (update headers
+                                                                 (string/lower-case (.getName header))
+                                                                 (fn [v]
+                                                                   (str (when v
+                                                                          (str v ","))
+                                                                        (.getValue header)))))
+                                                       {}
+                                                       (.getAllHeaders http-request))
+                                       ring-req-map {:method       method
```

Comment:
This is bad.
Half of the work of turning  "apache request" into "ring request" you do inside doExecute
Another half inside `->ring-request`

your `doExecute` should look like this
```
(doExecute [... ]
   (let [ring-request (->ring-request ...)
           ring-response (ring-handler ring-response)]
         .....))
```